### ISSUE TITLE
Tracking issue: Releases

### ISSUE THREAD

**Issue by jprochazk** [rerun-io/rerun#2530]
- [x] https://github.com/rerun-io/rerun/issues/1343
- [x] https://github.com/rerun-io/rerun/issues/2234
- [x] Research: How much of the release can we automate?
  - We can automate most things. What we can't (and/or don't want) to automate is:
    - Updating `CHANGELOG.md`
    - Creating the GitHub release
    - Social media posts
- [x] Implement new release process + Versioning scheme
  - https://github.com/rerun-io/rerun/pull/2961
  - https://github.com/rerun-io/rerun/pull/2973
  - https://github.com/rerun-io/rerun/pull/2979
  - https://github.com/rerun-io/rerun/pull/2984
  - https://github.com/rerun-io/rerun/pull/2992
  - https://github.com/rerun-io/rerun/pull/2995
  - https://github.com/rerun-io/rerun/pull/2998
  - https://github.com/rerun-io/rerun/pull/3000
- [ ] Update GCS load balancers to use `/version/latest` instead of hardcoded version
- [ ] Delete `latest` tag in favor of a branch that will be force-pushed on every release
- [ ] Test the new release process on an actual release

**Comment by jleibs** [comment]
We've already run into quite a few issues with having a `latest` tag. TLDR: git does not like re-tagging: https://git-scm.com/docs/git-tag#_on_re_tagging  if we want to have a convenience-pointer to "the most recent release thing" we should use a branch.

**Comment by abey79** [comment]
My 2ct:
- (1) is a bit overkill imho. I think the release process could be:
  - Manually ensure the version is correct (or bump it)—a script could be useful for that.
  - Manually finalise the release notes.
  - Commit what precedes.
  - Tag the commit with a well defined short semver pattern `0.8.0`.
- the process of creating a RC and a final release should be unify and do the right thing based on the (`0.8.0` or `0.8.0rc1`). AFAIK, the "right thing" is done automatically by PyPI (a `0.8.0rc1` version is considered pre-released and requires `pip install --pre`), so nothing specific to be done on our side.

Corollaries of that:
- No branch would be created automatically. As a matter of fact, I'm not being very prescriptive on the branching pattern. The release workflow should be triggered based on tag, regardless of where that tag is. For example, one could create a version bump/changelog commit on `main`, tag it for the initial beta/rc, then create a `releases/0.8` branch off that commit for further fix/RCs/etc.
- One nice convention we could have is to release alphas whenever we feel like it directly off `main`, and release RCs off a ?`release/0.8` branch when we are in the final test phase shortly before the final release.
- The version tag must match the version set in `Cargo.toml`/`pyproject.toml`/etc. I dislike this redundancy and it would be nice for the release workflow to check consistency there.

FYI, hear the release workflow for my main python project. Note in particular the "patterned tag" trigger: https://github.com/abey79/vpype/blob/master/.github/workflows/release.yml

**Comment by abey79** [comment]
@jleibs As far as wheels/crates are concerned, I'd just release a train of alphas off `main` as noted above. For landing, I'm not sure.

**Comment by jleibs** [comment]
> The release workflow should be triggered based on tag, regardless of where that tag is.

This used to be the case and I found it to be quite painful, which is what led us to transition from "tag drives release" to "release drives tag".

The fundamental problem is that at the time you are creating the tag, you often don't know whether or not the tree is in a good state, and the only way to find out is to create the tag. But then, in many situations we would tag a release only to then discover some aspect of the release was broken ranging from the build/CI tooling itself to the packages. This then meant either orphaning entire versions that had been prematurely tagged or going through heavy-handed force-retagging operations that are error prone and leave git sad.

Having the tag applied at the end, only upon successful completion of build and testing, means that if things fail you never bother tagging a bad commit.

**Comment by jprochazk** [comment]
The consequence of doing (1) manually is that we now need consistency checks. It's much simpler to have the workflow do the right thing automatically. It also doesn't remove the possibility of releasing often, the created branches would be based on `main` every time. I also don't think it's overkill, we're just moving those few manually executed commands into the workflow. Having the workflow create the branch would also ensure we can perform builds/tests *before* creating branches so that we don't accidentally produce a bunch of garbage in the git repository.

One important thing is that we should never commit into the release branches. They should be treated like the `gh-pages` branch - "automatically managed" by our tooling. That's maybe the core of my idea here: Releases are fully automatic; we only test them (because we don't have any real testing infrastructure in place)

**Comment by jleibs** [comment]
Something worth discussing a bit more might be the concepts of "alpha" vs "rc". We are currently using "alpha" releases more as RCs, but to @abey79 's point it could be quite helpful to more frequently (maybe weekly) produce alpha builds for continuous testing.

When it comes to the *actual* release I find that releasing from main is practically very challenging. The only way to do it successfully is to freeze all development and have the entire team switch to "release" mode (bug-fix only, no new features) until after the release is cut. Most folks find this to be disruptive enough to warrant cutting a release branch at the beginning of each release cycle with the "bug-fix-to-main + cherry-pick" workflow.  This is where the current recommendation to create a `release-x.y` branch at the beginning of each cycle comes from.

This is what I would imagine a healthy release history might look like:
```
        main 
         |
         | 0.9-alpha1
         |
         | 0.9-alpha2
         |
         |---  release-0.9 (branch)
unstable |    \ 
feature  |     | 0.9.0-rc1
         |     |
 bugfix  | --> | 0.9.0-rc2
         |     |
         |     * 0.9.0
         |
```

**Comment by abey79** [comment]
@jleibs I see your point indeed. I have been successfully force retagging, but never in a collaborative environment—I see how this could wrong. I don't have a fundamental issue against the "release drives tag" workflow, really.

<br/>

> One important thing is that we should never commit into the release branches. They should be treated like the gh-pages branch - "automatically managed" by our tooling. That's maybe the core of my idea here: Releases are fully automatic; we only test them (because we don't have any real testing infrastructure in place).

@jprochazk How would you deal with patch release? It's pretty nice to be able to commit/cherrypick a few commits in the `release/0.8` branch (post `0.8.0` release) towards a `0.8.1` patch release.

<br/>

Another important note: when release an alpha off `main`, e.g. the current `0.8.0a0`, it's import that `main` should be immediately bumped to `0.8.0.a1` to avoid pip refusing to install from `main` after `pip install rerun-sdk==0.8.0a0` on ground that the "current" version is already installed in the venv.

**Comment by abey79** [comment]
> Something worth discussing a bit more might be the concepts of "alpha" vs "rc". We are currently using "alpha" releases more as RCs, but to @abey79 's point it could be quite helpful to more frequently (maybe weekly) produce alpha builds for continuous testing.

Yes!

> ```
>         main 
>          |
>          | 0.9-alpha1
>          |
>          | 0.9-alpha2
>          |
>          |---  release-0.9 (branch)
> unstable |    \ 
> feature  |     | 0.9.0-rc1
>          |     |
>  bugfix  | --> | 0.9.0-rc2
>          |     |
>          |     * 0.9.0
>          |
> ```

I have this exact scheme in mind indeed.

**Comment by jprochazk** [comment]
@jleibs

> have the entire team switch to "release" mode (bug-fix only, no new features) until after the release is cut

I think this is fine as long as this "release mode" doesn't last too long. Any development happening in parallel with the release could stay in its own branch/PR for the duration of "release mode". Some things that will help with that:

- Require PRs to sign off on having tested their PR with all web examples (with exceptions for stuff that is guaranteed to not have any direct effect on `rerun` itself)
- Testing more often in general, and increasing the scope of automated testing
- Small but frequent alpha/rc releases

This way we'll hopefully end up with fewer bug fixes during "release mode", shortening that window and making it less disruptive to our trunk-based development workflow.

But the "have release branches be fully managed" is more of a strong suggestion, in rare cases we could manually create the release branch and cherry-pick bug fixes from `main` into it. The bulk of the work should still be done by CI in this case, so the workflows should be split up to enable that.

`pre-release checks` + `rc branch` + `publish rc` could be three reusable workflows wrapped into one which we use for the "happy" (and hopefully common) path.

@abey79

> How would you deal with patch releases?

The same way as any other release. If they can't be based off of `main` directly (which should be rare), create a new branch off of a good commit (maybe the previous CI-generated release rc branch), cherry-pick commits into it, then run the workflow on that - it will generate an `rc` branch and everything will be managed as usual from that point on.

To support that, (1) also needs to accept a base branch as input.

**Comment by jprochazk** [comment]
What I'm trying to say: I think instead of looking for ways to work around the fact that we can't release `main` easily, we should try to make `main` releasable (and only in rare cases not releasable)

**Comment by jprochazk** [comment]
Feature-flagging is another option for "things which live on main but aren't yet supposed to be released"

**Comment by abey79** [comment]
> What I'm trying to say: I think instead of looking for ways to work around the fact that we can't release `main` easily, we should try to make `main` releasable (and only in rare cases not releasable)

I think we have to be realistic and accept that our style of direct-to-`main`, fast-action, small PRs means that the large majority of commits on `main` will always be "not a good place to start a release branch". I don't think this is a problem BTW—personally, I like it a lot that way. Also, I dislike having any constraint on the team when a release is in progress (even if it might work now, it won't scale as we add engineers in the team).

These are the reason why I like @jleibs's proposed branch pattern, and am in favour of a release process that accommodates it.

<br/>

My current understanding is that the motivation for having the release workflow creating a branch for each release is that it wants to create a commit (with version bump, etc.) and want to ensure no "human" activity conflicts with this. Is that correct? Is there another motivation?

**Comment by jprochazk** [comment]
Yes, that's the motivation. If the branch from which the release candidate is created is in a good state, then the release candidate branch will also always be in a good state. Having a human-managed release branch means performing consistency checks, which are inherently fragile.

I started with a similar idea as the cherry-picked release branch (because that's what we did since I started 2 months ago), but during today's standup, someone (I don't remember who, maybe Emil?) said that a cherry-picked release branch was a bandaid for the last 2 releases, and that we shouldn't head in that direction, and I completely agree with that.
It's super fragile, because you can't always easily cherry-pick things into the release branch. You may end up with merge conflicts, which introduces an extra layer of "human mistake" potential, because you could introduce a bug by fixing the conflict.

Not only have we had to cherry-pick bug fixes (and other stuff) into release branches in the last 2 releases, but we've also had to cherry-pick stuff *out of the release branch*. That takes a lot of time.
From what I understand, the team is going to stay small for the foreseeable future, so in my opinion, scaling is not much of a concern yet, and can be addressed when rerun starts experiencing growing pains. Before then it's premature optimization. What is a concern is having a small team spend super valuable time doing tasks which should be (and can be) automated.

Having the release branches be fully managed means we're incentivized to just fix `main` and keep it in a good state *when trying to do a release*. It doesn't always have to be in a good state. If `main` is not in a good state to release, we will find that out when we attempt to make a release (or before then, when we realise we have unreleasable features on `main`).
In that case, we push the necessary bug fixes directly to main (or `#[cfg]` the unreleasable stuff), then easily create a new release branch. The work that we'd do in a `release` branch we'd instead do in `main`.

In short: We can continue to push small PRs directly to main. The only constraint we place on ourselves is to fix `main` before trying to do a release.

**Comment by jprochazk** [comment]
I'll try to summarize our positions:
A) longer-lived human-managed release branches
B) short-lived CI-managed release branches

For (A), we would create a release-x.y.z branch and use a script to bump versions. Then we'd cherry-pick commits into it for bug fixes or specific features.

For (B), the branch would be release-x.y.z-rcN (release candidate), and release-x.y.z (full release). It would always be based on main, so any bug fixes would go to main, and any not-quite-finished features would have to be #[cfg]'d away.

(B) would require introducing the constraint that we get main into a good state before doing a release, as opposed to choosing a good commit in main, then cherry-picking other stuff into it. It would mean that for each release, we'd enter into a "release mode" where any new features will have to wait until the release is done. It's not a complete halt on feature development, but we'd have to make sure unfinished features don't end up in the release.

In both cases, a lot of the currently manual work would be automated (bumping verisons, releasing docs and publishing crates/wheels). I'm not totally dead-set on (B), I just think it's less fragile, and forces us to do things which we'd have to do anyway.

**Comment by abey79** [comment]
I understand your arguments (including that "premature optimisation" point 👍🏻). Ultimately, both approaches can work—there is an element of philosophy/taste in that decision.

Small correction though: for (1), the branch is `release-x.y` and it contains multiple tags `x.y.0`, `x.y.1`, etc. (plus all the RCs of course). This basically "bakes in" the notion that patch release are exactly that: a (small) patch over the last minor release.

**Comment by abey79** [comment]
(If we're discussing taste, I'd rather have `releases/x.y` than `release-x.y`, as git UIs can exploit that hierarchy, e.g. to collapse all release branches in the list.)

**Comment by jprochazk** [comment]
For now I'll focus on getting all the underlying stuff automated, so we could choose either approach at some point. Would love to get more opinions from the rest of the team here, though!

**Comment by jleibs** [comment]
Regarding the last 2 releases, they have both been fairly significant long-lived branches and I believe this is what @emilk was referring to as atypical.  0.6 was cut a few weeks in advance of the actual release and 0.7 was 0.6 + cherry-picks.  I think we're all in agreement that neither of those was an optimal workflow.

I think a good indicator is whether we are cherry-picking *features* from main into the branch. If are are cherry-picking anything other than a small number of bug-fixes, it's a strong signal that we branched too early. 

Also, I'll point out that if the only development work happening is bug-fixes for the release, then every commit to main gets ff-merged into release-x.y and they just stay in lockstep, which is the same as releasing from main :shrug:

**Comment by emilk** [comment]
Let's make sure the new release work-flow is properly documented in `RELEASES.md`

**Comment by jprochazk** [comment]
My original idea here is too inflexible. There is currently no "average" release, so trying to "box" the release process is not going to work, at least not yet. It's important that we can manually edit the release PR and not waste too much time on that. I think there's still a lot of room for improvement in the release process, and I'll update the OP with a different checklist that would hopefully better accomodate our needs.
